### PR TITLE
feat: add a sparse-checkout workflow

### PR DESCRIPTION
**Pull request by raisedadead**
This PR aims to be the most non-intrusive approach to manage getting the files that one needs for working on the platform.

The strategy is as follows:

- We have close to 90K files related to curriculum.
- A fresh clone, of the repo (before this PR lands) has these stats:
  
  ```shell
  ❯ git clone https://github.com/freeCodeCamp/freeCodeCamp.git
  ```

  ```shell
  ❯ git count-objects -vH
	count: 0
	size: 0 bytes
	in-pack: 551871
	packs: 2
	size-pack: 438.05 MiB
	prune-packable: 0
	garbage: 0
	size-garbage: 0 bytes
  ```

- We could instead use `--sparse` and `--filter=blob:none` to get only git related information.

  ```shell
	❯ git clone --filter=blob:none --sparse https://github.com/freeCodeCamp/freeCodeCamp.git
  ```
  ```shell
	❯ git count-objects -vH
	count: 0
	size: 0 bytes
	in-pack: 249102
	packs: 3
	size-pack: 59.60 MiB
	prune-packable: 0
	garbage: 0
	size-garbage: 0 bytes
  ```

- Now to get the actual files in the filesystem we would do something along the lines of 

  ```bash
  git sparse-checkout init
  git sparse add api
  git sparse add client
  ...
  git checkout main
  ```
- However doing the above manually is tedious and error prone. Instead this PR adds a bash script to help. There are detailed commands and steps are in the comments in the file.

  Critically, downloading the needed directories:

  ```shell
	❯ ./bootstrap.sh --checkout english
	 :> Checking out language folders: english...
	 :> Initializing sparse-checkout with base configuration...
	remote: Enumerating objects: 229, done.
	remote: Counting objects: 100% (113/113), done.
	remote: Compressing objects: 100% (48/48), done.
	remote: Total 229 (delta 87), reused 65 (delta 65), pack-reused 116
	Receiving objects: 100% (229/229), 170.02 KiB | 1.16 MiB/s, done.
	Resolving deltas: 100% (133/133), done.
	 :> Sparse-checkout initialized.
	remote: Enumerating objects: 7288, done.
	remote: Counting objects: 100% (5740/5740), done.
	remote: Compressing objects: 100% (4784/4784), done.
	remote: Total 7288 (delta 1310), reused 956 (delta 956), pack-reused 1548
	Receiving objects: 100% (7288/7288), 6.21 MiB | 7.00 MiB/s, done.
	Resolving deltas: 100% (1853/1853), done.
	Updating files: 100% (7288/7288), done.
	 :> Checkout of specified languages completed.
  ```

  reduces the repos size on the local machine:

  ```shell
	❯ git count-objects -vH
	count: 0
	size: 0 bytes
	in-pack: 256619
	packs: 5
	size-pack: 66.18 MiB
	prune-packable: 0
	garbage: 0
	size-garbage: 0 bytes
  ```

  and the count of files, for example:

  ```shell
	❯ cd curriculum/challenges/english
	❯ find . -type f | wc -l
	    7286
  ```

> [!NOTE] 
> This PR aims to have minimal impact.

- This PR should will not affect any existing flows that we have like CI & CD.
- The idea is to add instructions for using this script and workflows to our documentation. - Contributors can simply opt-in to follow this workflow to grab only the needed parts of the repository.
- Follow-up PRs may optimize some automations like CI, etc. but that is left to discretion of the dev-team.

### CODE REVIEW

**Comment by raisedadead** [comment]
This is more like a demo PR. 

I will have to adjust some nuances around files and their location. Example there is a `.markdownlint.yaml` buried in the challenges folder for some reason and so on.

**Comment by raisedadead** [comment]
For reviewers:

Follow the steps in the `bootstrap.sh` file, replacing my fork `https://github.com/raisedadead/freeCodeCamp.git` for testing.

Ideally you want to clone from scratch to avoid overwriting any existing setups you have.

File: bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
+
+# Usage instructions for cloning and setting up sparse-checkout
+# 1. Clone the repository sparsely:
+#    git clone --filter=blob:none --sparse https://github.com/freeCodeCamp/freeCodeCamp.git
```
Comment (ShaunSHamilton) [inline bootstrap.sh:5]: This helps me a lot.

```suggestion
#    git clone --depth=1 --filter=blob:none --sparse https://github.com/freeCodeCamp/freeCodeCamp.git
```

Comment (raisedadead) [reply bootstrap.sh:5]: The issue with shallow clones is it is great for CI/CD (even one-off branches like on Gitpod and Codespaces) but rarely works in a localized setup.

In a local setup, you are very likely to switch branches, rebase, etc. which become buggy or non-functional at all.

I am aiming for only the first time setup (with incremental get me more files) for now.

**Review by moT01 — COMMENTED** [review]
This is pretty neat. I had to checkout the `_meta` folder to be able to build - `git sparse-checkout add curriculum/challenges/_meta`. Perhaps that could be added to the script.

**Review by ojeytonwilliams — COMMENTED** [review]
How about we do something like nvm and ask contributors to curl the script? That way the user has to copy something like

```bash
curl https://raw.githubusercontent.com/freeCodeCamp/freeCodeCamp/main/bootstrap.sh | bash
```

for the initial setup. If we fold 

```bash
git clone --filter=blob:none --sparse https://github.com/freeCodeCamp/freeCodeCamp.git
cd freeCodeCamp
git sparse-checkout init --no-cone
```

into the script and default to checking out english, then the whole process is nice and simple.

**Pull request closed without merging**